File: apps/api/app/models/budget.py
import uuid
import enum
from datetime import datetime
from sqlalchemy import String, Numeric, Boolean, ForeignKey, DateTime, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func
from app.database import Base


class BudgetPeriod(str, enum.Enum):
    daily     = "daily"
    weekly    = "weekly"
    monthly   = "monthly"
    quarterly = "quarterly"
    annual    = "annual"


class Budget(Base):
    __tablename__ = "budgets"

    id:          Mapped[str]          = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id:     Mapped[str]          = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), index=True)
    category_id: Mapped[str]          = mapped_column(ForeignKey("categories.id"), index=True)
    amount:      Mapped[float]        = mapped_column(Numeric(14, 2), nullable=False)
    currency:    Mapped[str]          = mapped_column(String(3), nullable=False)
    period:      Mapped[BudgetPeriod] = mapped_column(SAEnum(BudgetPeriod), default=BudgetPeriod.monthly)
    is_active:   Mapped[bool]         = mapped_column(Boolean, default=True)
    created_at:  Mapped[datetime]     = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at:  Mapped[datetime]     = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    user     = relationship("User",     back_populates="budgets")
    category = relationship("Category", back_populates="budgets")

File: apps/mobile/src/db/models/Budget.ts
import { Model } from '@nozbe/watermelondb';
import { field, readonly, date, text } from '@nozbe/watermelondb/decorators';

export class Budget extends Model {
  static table = 'budgets';

  @text('user_id')     userId!: string;
  @text('category_id') categoryId!: string;
  @field('amount')     amount!: number;
  @text('currency')    currency!: string;
  @text('period')      period!: 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'annual';
  @field('is_active')  isActive!: boolean;
  @field('is_synced')  isSynced!: boolean;
  @text('server_id')   serverId!: string;

  @readonly @date('created_at') createdAt!: Date;
  @readonly @date('updated_at') updatedAt!: Date;
}
